pkg/syncer/storage_policy_quota_reconciler.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	cnsoperatorv1alpha1 "sigs.k8s.io/vsphere-csi-driver/v3/pkg/apis/cnsoperator"
+	storagepolicyusagev1alpha1 "sigs.k8s.io/vsphere-csi-driver/v3/pkg/apis/cnsoperator/storagepolicy/v1alpha1"
+	"sigs.k8s.io/vsphere-csi-driver/v3/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/v3/pkg/kubernetes"
+)
+
+// StoragePolicyQuotaReconciler is the interface for StoragePolicyQuota reconciler.
+type StoragePolicyQuotaReconciler interface {
+	// Run starts the reconciler.
+	Run(ctx context.Context, workers int)
+}
+
+type storagePolicyQuotaReconciler struct {
+	// metadataSyncer informer to get FSS information.
+	metadataSyncerInformer *metadataSyncInformer
+	// StoragePolicyQuota queue to add StoragePolicyUsage CRs.
+	svcQuotaAddQueue workqueue.RateLimitingInterface
+	// StoragePolicyQuota queue to delete StoragePolicyUsage CRs.
+	svcQuotaDelQueue workqueue.RateLimitingInterface
+	// StoragePolicyQuota Synced.
+	svcQuotaSynced cache.InformerSynced
+}
+
+// newStoragePolicyQuotaReconciler returns a StoragePolicyQuotaReconciler.
+func newStoragePolicyQuotaReconciler(
+	ctx context.Context,
+	resyncPeriod time.Duration,
+	metadataSyncerInformer *metadataSyncInformer,
+	svcStoragePolicyQuotaRateLimiter workqueue.RateLimiter,
+	stopCh <-chan struct{}) (*storagePolicyQuotaReconciler, error) {
+
+	log := logger.GetLogger(ctx)
+	// Create an informer for StoragePolicyQuota instances.
+	k8sConfig, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		return nil, logger.LogNewErrorf(log, "failed to get kubeconfig with error: %v", err)
+	}
+	dynamicInformerFactory, err := k8s.NewDynamicInformerFactory(ctx, k8sConfig, metav1.NamespaceAll, resyncPeriod, true)
```

Comment:
I think setting the resyncPeriod doesn't do anything atm. As far as I know it will lead to Update events every resyncPeriod for all objects, but we don't add an UpdateFunc

Reply:
I think after resyncPeriod time, all event in the informer cache are replayed, which can include add/delete events, hence kept the timeout.
Please let me know if add/delete does not get handled as stated above

Reply:
resync does roughly the following:
* list all objects in the local store
* enqueue the objects

I tried it locally and I got these events as "update" events.

I think you can easily test with adding some log statements in an Add & Update func


(it definitely does not replay past events)